libs/datamodel/core/src/transform/ast_to_dml/db/attributes.rs

```diff
@@ -106,6 +109,37 @@ pub(super) fn resolve_model_and_field_attributes<'ast>(
     autoincrement::validate_auto_increment(model_id, ctx);
 }
 
+pub(super) fn validate_index_names(ctx: &mut Context<'_>) {
+    if ctx.db.active_connector().supports_multiple_indexes_with_same_name() {
+        return;
+    }
+
+    let mut index_names = HashSet::new();
+    let mut errors = Vec::new();
+
+    for (attr, index) in ctx
+        .db
+        .ast
+        .iter_models()
+        .flat_map(|(model_id, _)| ctx.db.types.models[&model_id].indexes.iter())
+    {
+        let index_name: &str = match index.name {
+            Some(index_name) => index_name,
+            None => continue,
+        };
+
+        if index_names.insert(index_name) {
+            continue; // true means this name hasn't been seen before
+        }
+
+        errors.push(DatamodelError::new_multiple_indexes_with_same_name_are_not_supported(
+            index_name, attr.span,
+        ))
+    }
+
+    errors.into_iter().for_each(|err| ctx.push_error(err))
```

Comment:
This reimplements the existing validation, but I suspect it should validate a lot more.